lib/cursor.js

```diff
@@ -768,6 +769,29 @@ class Cursor extends CoreCursor {
     }
   }
 
+  /**
+   * Set the ReadPreference for the cursor.
+   * @method
+   * @param {(string|ReadConcern)} readConcern The new read preference for the cursor.
+   * @throws {MongoError}
+   * @return {Cursor}
+   */
+  setReadConcern(readConcern) {
+    if (this.s.state !== CursorState.INIT) {
+      throw new MongoError('cannot change cursor readConcern after cursor has been accessed');
+    }
+
+    if (readConcern instanceof ReadConcern) {
+      this.options.readConcern = readConcern;
+    } else if (typeof readConcern === 'string') {
+      this.options.readConcern = new ReadConcern(readConcern);
+    } else {
+      throw new TypeError('Invalid read concern: ' + readConcern);
+    }
```

Comment:
I wonder if this logic belongs on the `ReadConcern` class itself, perhaps `ReadConcern.resolve(readConcernLike)`? It looks like `ReadConcern.fromOptions` does a similar check

Reply:
We've already handled this in master and have pretty good `fromOptions` methods there, I can port those over, I just thought we'd be anti the back-porting. [In the master branch I re-wrote this to use `.fromOptions` for both `setReadPreference` and `setReadConcern`.](https://github.com/mongodb/node-mongodb-native/pull/2584/files#diff-f7a848a5828483a0958ec4e436ca7f277d846a669c610b5647e192450331d09dR754-R773)

Let me know if you're interested in a back-port of those classes.

Reply:
done, backported updated .fromOptions from both classes and have these methods same as master.

Reply:
Ok, I think it's a good start to backport the `fromOptions` changes, but I'm still wondering if we can improve things a bit. It seems like generally in both `ReadConcern.fromOptions` and this new method that we want to be able to take something that is `ReadConcernLike`, convert it to a `ReadConcern` instance, and throw an error if the provided value doesn't satisfy the requirements of `ReadConcernLike`. It seems to me like an intermediate method could be used in both places (I suggested `resolve` above), and if we had that method we could avoid the extra allocation here when passing an object to `fromOptions` in this method.

Reply:
I've gone down a rabbit hole to refactor and fix `ReadConcern` and `ReadPreference`. ReadPreference seems pretty buggy and there's a lot of duplication. And I also would love to replace `ReadPreference.translate` and use a `options = resolve()` in-place. This request here has a bit of scope-creep which is why I first implemented it the way I did. Complications arise when deciding default values, validation, returning undefined / throwing / default, and parental inheritance, and some type guard functions. I kind of have an opinionated way of how all this should be done. Also I think there's a bit of overlap with Neal's client options work as well.  I'm not sure what to do here, I don't want to respond with an unexpected massive refactor of these classes. Part of me just wants to rip the band-aid off and fix them.